clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -397,6 +398,47 @@ private synchronized void requestUpdateForNewTopics() {
         requestUpdate();
     }
 
+    private void validateCluster(Cluster newCluster) {
+        String previousClusterId = this.cluster.clusterResource().clusterId();
+        String newClusterId = newCluster.clusterResource().clusterId();
+
+        if (previousClusterId != null && newClusterId != null && !previousClusterId.equals(newClusterId)) {
+            // kafka cluster id is unique.
+            // On client side, the cluster id in Metadata is only null during bootstrap, and client is
+            // expected to talk to the same cluster during its life cycle. Therefore if cluster id changes
+            // during metadata update, meaning this metadata update response is from a different cluster,
+            // client should reject this response and not update cached cluster to the wrong cluster
+
+            // Since removing brokers and adding to another cluster can be common operation,
+            // it might be more suitable to just throw an exception for update and fail this update operation
+            // instead of bringing down the client completely, so that the metadata can be updated later from
+            // other brokers in the same cluster.
+
+            // remove the traitor broker from the cluster's node list
+            List<Node> originalNodes = new ArrayList<>(this.cluster.nodes());
+            for (Node node : newCluster.nodes()) {
+                originalNodes.remove(node);
```

Comment:
The node hash code is computed on (id, host, port, rack).  If the node is moved to a different rack in a different cluster, this doesn't clean up stale node?